src/Integration.Vsix.UnitTests/CFamily/CLangAnalyzerTests.cs

```diff
@@ -29,21 +33,110 @@ namespace SonarLint.VisualStudio.Integration.Vsix.CFamily.UnitTests
+        }
+
+        [TestMethod]
+        public void ExecuteAnalysis_MissingProjectItem_NoAnalysis()
+        {
+            serviceProviderWithValidProjectItem = CreateServiceProviderReturningProjectItem(null);
+
+            var testSubject = new TestableCLangAnalyzer(telemetryManagerMock.Object, new ConfigurableSonarLintSettings(),
+                rulesConfigProviderMock.Object, serviceProviderWithValidProjectItem.Object, testLogger);
+
+            testSubject.ExecuteAnalysis("path", "charset", new[] { AnalysisLanguage.CFamily }, Mock.Of<IIssueConsumer>(), null, CancellationToken.None);
+
+            testSubject.CreateRequestCallCount.Should().Be(0);
+            testSubject.TriggerAnalysisCallCount.Should().Be(0);
+        }
+
+        [TestMethod]
+        public void ExecuteAnalysis_ValidProjectItem_RequestCannotBeCreated_NoAnalysis()
+        {
+            var testSubject = new TestableCLangAnalyzer(telemetryManagerMock.Object, new ConfigurableSonarLintSettings(),
+                rulesConfigProviderMock.Object, serviceProviderWithValidProjectItem.Object, testLogger);
+            testSubject.RequestToReturn = null;
+
+            testSubject.ExecuteAnalysis("path", "charset", new[] { AnalysisLanguage.CFamily }, Mock.Of<IIssueConsumer>(), null, CancellationToken.None);
+
+            testSubject.CreateRequestCallCount.Should().Be(1);
+            testSubject.TriggerAnalysisCallCount.Should().Be(0);
+        }
+
+        [TestMethod]
+        public void ExecuteAnalysis_ValidProjectItem_RequestCanBeCreated_AnalysisIsTriggered()
+        {
+            var testSubject = new TestableCLangAnalyzer(telemetryManagerMock.Object, new ConfigurableSonarLintSettings(),
+                rulesConfigProviderMock.Object, serviceProviderWithValidProjectItem.Object, testLogger);
+            testSubject.RequestToReturn = new Request();
+
+            testSubject.ExecuteAnalysis("path", "charset", new[] { AnalysisLanguage.CFamily }, Mock.Of<IIssueConsumer>(), null, CancellationToken.None);
+
+            testSubject.CreateRequestCallCount.Should().Be(1);
+            testSubject.TriggerAnalysisCallCount.Should().Be(1);
+        }
+
+        private static Mock<IServiceProvider> CreateServiceProviderReturningProjectItem(ProjectItem projectItemToReturn)
+        {
+            var mockSolution = new Mock<Solution>();
+            mockSolution.Setup(s => s.FindProjectItem(It.IsAny<string>())).Returns(projectItemToReturn);
+            var solution = mockSolution.Object;
+
+            var mockDTE = new Mock<DTE>();
+            mockDTE.Setup(d => d.Solution).Returns(solution);
+            var dte = mockDTE.Object;
+
+            var mockServiceProvider = new Mock<IServiceProvider>();
+            mockServiceProvider.Setup(s => s.GetService(typeof(DTE))).Returns(dte);
+
+            return mockServiceProvider;
+        }
+
+        private class TestableCLangAnalyzer : CLangAnalyzer
```

Comment:
I understand why you did it this way, and I saw the comment that you wanted to avoid creating new interfaces.

I think it's worth creating those interfaces.
In general, I'm not a fan of the internal methods that we expose only for testing. It disrupts code flow, and almost always ends with flow paths that are never tested, since we are laser-focused on testing just the internal method and not the whole class flow. It always means that there is a problem with the responsibility of the class, which should be resolved by breaking it down to smaller components.

In this particular case, an interface would be much simpler to read, test and maintain, and could easily be mocked, instead of inheriting from it.

Reply:
Let's discuss this offline. I don't think it's worth spending the time on now; it wasn't the purpose of this PR. This class needs more substantial refactoring to be properly testable, and probably to fix the threading issue too.